### PR TITLE
Quantum RPG Battle Engine

### PR DESCRIPTION
**Pull request by dstrain115**
- Add class to actually run a Quantum RPG battle.
- This pits players versus NPCs in a battle.
- Player actions and NPC actions can be triggered via functions.
- Also adds a simple test NPC.

At this point, the MVP of the RPG is almost playable, though some work is needed to initialize and run the battle (plus make it more interesting).

### CODE REVIEW

File: unitary/examples/quantum_rpg/npcs.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    """
+
+    def is_npc(self):
+        return True
+
+
+class Observer(Npc):
+    """Simple test NPC that measures a random qubit each turn."""
+
+    def is_npc(self):
```
Comment (losos0) [inline unitary/examples/quantum_rpg/npcs.py:19]: This override is not needed

File: unitary/examples/quantum_rpg/battle.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+        1) Retrieve the possible actions from the player.
+        2) Prompt the player for which action to use.
+        3) Prompt the player which NPC and qubit to target.
+        4) Call the player's action to perform the action.
+
+        Args:
+            user_input: List of strings that substitute for the user's
+                raw input.
+        """
+        idx = 0
```
Comment (losos0) [inline unitary/examples/quantum_rpg/battle.py:89]: How about
```
if user_input is not None:
   user_input = iter(user_input)
   get_user_input = lambda unused_prompt: next(user_input)
else:
   get_user_input = input
```

File: unitary/examples/quantum_rpg/npcs.py (hunk 1 above)
Comment (dstrain115) [reply unitary/examples/quantum_rpg/npcs.py]: Removed.

File: unitary/examples/quantum_rpg/battle.py (hunk 2 above)
Comment (dstrain115) [reply unitary/examples/quantum_rpg/battle.py]: Great idea, very elegant.  Done.

File: unitary/examples/quantum_rpg/battle.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            print(f'{current_player.name} turn:', file=self.file)
+            if not current_player.is_active():
+                print(f'{current_player.name} is DOWN!', file=self.file)
+                continue
+            actions = current_player.actions()
+            for key in actions:
+                print(key, file=self.file)
+            action = get_user_input('Choose your action: ')
+            if action in current_player.actions():
+                monster = int( get_user_input('Which enemy number: ')) + 1
```
Comment (smitsanghavi) [inline unitary/examples/quantum_rpg/battle.py:98]: Shouldn't this be subtracting 1 from the input?

File: unitary/examples/quantum_rpg/battle.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+                if monster < len(self.enemy_side):
```
Comment (smitsanghavi) [inline unitary/examples/quantum_rpg/battle.py:99]: Print an error when an invalid monster number is provided

**Review by smitsanghavi — CHANGES_REQUESTED** [review]

File: unitary/examples/quantum_rpg/battle.py (hunk 3 above)
Comment (dstrain115) [reply unitary/examples/quantum_rpg/battle.py]: Good catch!  Fixed and added more unit tests for this case.

File: unitary/examples/quantum_rpg/battle.py (hunk 4 above)
Comment (dstrain115) [reply unitary/examples/quantum_rpg/battle.py:99]: Done (and tested more thoroughly)

**Comment by dstrain115** [comment]
Thanks for the reviews everyone.  Probably last update before 2023, but hope to push on this some more next year!

**Review by smitsanghavi — APPROVED** [review]

**Comment by smitsanghavi** [comment]
Looks awesome, excited to explore what all can be done with this framing!

**Pull request merged**